### PR TITLE
HTTP 101 (Core protocol) updated

### PR DESCRIPTION
**Pull request by bverschueren**

### CODE REVIEW

**Comment by bverschueren** [comment]
Hi, any feedback on this ?

File: http_101.rst [hunk 1]
```diff
@@ -1,20 +1,369 @@
...
+PUT           Replace all current representations of the target resource with the requested payload.
+DELETE        Remove all current representations of the target resource.
+CONNECT       Establish a tunnel to the server identified by the target resource.
+OPTIONS       Describe the communication options for the target resource.
+TRACE         Perform a message loop-back test along the path to the target resource.
+============= ===============
+
+(source: :rfc:`7231`)
+
+After the empty line, an eventual body message contains the payload (data) of the request.
```
Comment (miketheman) [inline http_101.rst]: Can you clarify that the body is part of the response?

File: http_101.rst [hunk 2]
```diff
@@ -1,20 +1,369 @@
...
+
+Response
+========
+
+The server parses the response, and, depending on the validity of the request, sends one or more responses.
+An HTTP response starts with the status line, which contains the protocol version, succes or error code, and a reason message.
```
Comment (miketheman) [inline http_101.rst]: succes => success

File: http_101.rst [hunk 3]
```diff
@@ -1,20 +1,369 @@
...
+
+The status line is mostly followed by a header section containing server information and other metadata.
+An empty line indicates the end of the status line and header.
+The final part of the response, the message body, contains the payload send back to the client.
```
Comment (miketheman) [inline http_101.rst]: send => sent, or to send

File: http_101.rst [hunk 4]
```diff
@@ -1,20 +1,369 @@
...
-request
+Telnet
+------
+
+The telnet utility is used for interactive communication to a host on a given port. 
+Once the connection to the remote host is established, an HTTP request can be send to the host by typing it in the prompt:
+
+.. code-block:: console
+
+	$ telnet apache.org 80
```
Comment (miketheman) [inline http_101.rst]: Could you possibly try examples with opsschool.org? Could help readers try this out.

File: http_101.rst [hunk 5]
```diff
@@ -1,20 +1,369 @@
...
+	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+        	                         Dload  Upload   Total   Spent    Left  Speed
+	100  100M  100  100M    0     0   400M      0 --:--:-- --:--:-- --:--:--  414M
+
+When starting to download a large file with ``-C`` option, the download can be resumed in case of an interrupt.
+
+cURL can use a proxy to retrieve a resource when specified to do so using the ``-x`` or ``--proxy`` option with the proxy specified as ``[protocol://][user:password@]proxyhost[:port]``.
+
+By default cURL does not follow HTTP redirects, and instead a 3xx redirection message is given.
+The ``-L`` switch makes cURL automatically redo the request to the given location till it finds the targeted resource.
```
Comment (miketheman) [inline http_101.rst]: redo => follow redirects and issue another request...

File: http_101.rst [hunk 6]
```diff
@@ -1,20 +1,369 @@
...
+
+
+Then again in the first (server) terminal, the data is displayed:
+
+.. code-block:: console
+
+	$ cat output.log
+	I'm connected as client
+	
+Except for some simple pieces of text, netcat can be used to copy files and directory structures.
```
Comment (miketheman) [inline http_101.rst]: Except for some => Beyond sending

File: http_101.rst [hunk 7]
```diff
@@ -1,20 +1,369 @@
...
+
+cURL outputs a progress meter to the terminal, showing statistics about the operation.
+However, once cURL is about to write data to the terminal, the display of the progress meter is disabled so it may seem as if it was never present.
+The progress meter can still be seen when the output is written ``-o <file>`` or piped ``>`` to a file.
+
+In case the download takes some time (for example downloading a big file) the progress meter can still be seen in the terminal:
+
+.. code-block:: console
+
+	$ curl -O http://localhost/bigfile
```
Comment (adrianmoisey) [inline http_101.rst]: The text above uses `-o <file>`, but the example doesn't say what `-O` means. Is that fine?

File: http_101.rst [hunk 8]
```diff
@@ -1,20 +1,369 @@
...
+
+	<html>Hello there world!</html>
+	Connection closed by foreign host.
+
 
 Apache, nginx
 =============
 
+Apache and Nginx are the two most used open-source webservers on the internet.
+Though they share several features, architecturally they differ significally.
```
Comment (miketheman) [inline http_101.rst]: architecturally they differ significally => their architectures differ significantly.

File: http_101.rst [hunk 9]
```diff
@@ -1,20 +1,369 @@
...
+There are two mechanisms for implementing virtual hosts:
+
+IP-based
+--------
+
+In IP-based virtual hosting each virtual host is configured with its own dedicated IP address.
+The web server is configured to serve multiple IP addresses, either on multiple physical network interfaces or on virtualized interfaces on a single physical interface.
+Based on the IP on which a client connects, the web servers determines which of the configured virtual host to serve to the client. 
+
+Name based
```
Comment (adrianmoisey) [inline http_101.rst]: Should this be `Name-based`?

File: http_101.rst [hunk 10]
```diff
@@ -1,20 +1,369 @@
...
+Apache and Nginx are the two most used open-source webservers on the internet.
+Though they share several features, architecturally they differ significally.
+
+Apache relies on a process-based model, either with a single or multiple threads per process that handle a connection, whereas Nginx operates event-driven with multiple single-threaded processes that all handle connections asynchronously.
+Nginx' event-driven aproach requires less resources, making it faster, especially for static content.
+For dynamic content however, Nginx needs to proxy requests to an external processor for execution as it has no programming language support.
+Therefore Nginx is often used as a front-end/proxy server in combination with other application servers, possibly Apache.
+
+Both web servers are extensible with modules (security, compression, proxy, ...).
+Where Apache is able to dynamically load and enable modules at runtime, Nginx need to have additional modules compiled in the core to have it available.
```
Comment (miketheman) [inline http_101.rst]: Nginx need to have additional modules compiled in the core to have it available. 
=>
Nginx can only load modules that have been precompiled into the executable.

File: http_101.rst [hunk 11]
```diff
@@ -1,20 +1,369 @@
...
+Virtual hosting is a technique where mutiple websites are hosted on a single web server.
+The server's resources are shared amongst all configured hosts, but each host has its own IP address or domain name.
+There are two mechanisms for implementing virtual hosts:
+
+IP-based
+--------
+
+In IP-based virtual hosting each virtual host is configured with its own dedicated IP address.
+The web server is configured to serve multiple IP addresses, either on multiple physical network interfaces or on virtualized interfaces on a single physical interface.
+Based on the IP on which a client connects, the web servers determines which of the configured virtual host to serve to the client.
```
Comment (miketheman) [inline http_101.rst]: host => hosts

**Comment by miketheman** [comment]
@bverschueren awesome stuff, added some comments. Feel free to address them, and then squash commits down to a single commit in this PR.

File: http_101.rst [hunk 12]
```diff
@@ -1,20 +1,369 @@
...
+In IP-based virtual hosting each virtual host is configured with its own dedicated IP address.
+The web server is configured to serve multiple IP addresses, either on multiple physical network interfaces or on virtualized interfaces on a single physical interface.
+Based on the IP on which a client connects, the web servers determines which of the configured virtual host to serve to the client. 
+
+Name based
+----------
+
+In name-based virtual hosting a web server serves multiple host names on a single IP address.
+While a client is connecting to a webserver, the destination address is resolved to an IP address either by DNS or local lookup table.
+As several virtual hosts may be hosted on the same IP, the client needs to state which website it is reaching for by setting the ``Host`` header.
```
Comment (adrianmoisey) [inline http_101.rst]: what about "requesting" instead of "reaching for"?

**Comment by bverschueren** [comment]
Ok, reformulated some sentences (as a non-native English speaker my wording might be somewhat stiff :-) ) and some spelling mistakes (similarly, read over small spelling errors quite rapidly..)
Also changed the examples to target www.opsschool.org and extended the "cURL -L" redirection handling.

**Comment by bverschueren** [comment]
@miketheman @adrianmoisey I've addressed your comments and squashed into a single commit.
Any other remarks or you think it can be merged ?

File: http_101.rst [hunk 13]
```diff
@@ -1,20 +1,444 @@
...
+	Host: example.com
+
+
+The request line is followed by headers containing client information and other metadata.
+As per protocol specification, the client must include the ``Host`` header in the HTTP request.
+A single server may serve several (virtual) hosts, so the ``Host`` header signals the server to which of the possible hosts to route the request.
+Some implementations might interpret a request without ``Host`` header as invalid, and respond with the ``400 Bad Request`` message.
+An empty line indicates the end of the request line and header section.
+
+The available methods are :
```
Comment (adrianmoisey) [inline http_101.rst]: Extra space between "are" and ":"

**Comment by adrianmoisey** [comment]
1 small nitpick comment.
I think its fine, but I'm not a maintainer.

**Comment by miketheman** [comment]
@bverschueren admirable work! Going to merge, despite the extra space, as I _think_ sphinx will handle that. In case it doesn't, it's a nice PR for someone else to fix in the future. :)

**Comment by miketheman** [comment]
Thanks to @adrianmoisey for helping with review, too!

**Pull request merged**